File: app/src/main/java/com/example/proif_a01_java/View/ProductTilesFragments.java
package com.example.proif_a01_java.View;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import com.example.proif_a01_java.Model.Inventory;
import com.example.proif_a01_java.Model.Page;
import com.example.proif_a01_java.Model.Product;
import com.example.proif_a01_java.Presenter.ProductPresenter;
import com.example.proif_a01_java.databinding.FragmentProductTilesBinding;

import java.util.ArrayList;

public class ProductTilesFragments extends Fragment implements IProducts, View.OnClickListener {
    public Inventory inv;
    private FragmentProductTilesBinding binding;
    private FragmentManager fragmentManager;
    private ProductPresenter presenter;
    private ProductTilesAdapter adapter;
    private int availLoads;
    private int loadNumber;

    // constructor kosong
    public ProductTilesFragments () {

    }

    // singleton
    public static ProductTilesFragments newInstance(Inventory inv) {
        ProductTilesFragments fragments = new ProductTilesFragments();
        fragments.inv = inv;
        return fragments;
    }

    @Override
    public View onCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        this.fragmentManager = getParentFragmentManager();
        this.binding = FragmentProductTilesBinding.inflate(inflater, container, false);

        // inisiasi
        this.presenter = new ProductPresenter(this);
        this.adapter = new ProductTilesAdapter(getActivity(), this.fragmentManager);

        // set adapter
        this.binding.lvProducts.setAdapter(this.adapter);

        // filtering
        this.binding.etFilter.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                String search = binding.etFilter.getText().toString();
                presenter.searchProducts(search);
            }
        });

        // set onclick listener
        this.binding.ivList.setOnClickListener(this::onClick);
        this.binding.showMore.setOnClickListener(this::onClick);

        // load new products
        this.loadProducts();

        // set adapter
        this.binding.lvProducts.setAdapter(this.adapter);

        return this.binding.getRoot();
    }

    private void loadProducts () {
        ArrayList<Product> products = new ArrayList<>();

        for (int i=0; i<5; i++) {
            products.add(this.inv.products[i]);
        }

        this.loadNumber = 5;
        this.availLoads = this.inv.products.length - 5;

        this.presenter.loadProducts(products);
    }

    private void loadMoreProducts () {
        if (this.availLoads > 0) {
            ArrayList<Product> products = new ArrayList<>();
            int max;
            if (this.availLoads < 5) {
                max = this.loadNumber + this.availLoads;
            }else{
                max = this.loadNumber + 5;
            }
            for (int i=this.loadNumber; i<max; i++) {
                System.out.println(i);
                products.add(this.inv.products[i]);
            }
            this.presenter.addProducts(products);
            this.loadNumber = max;
            this.availLoads -= 5;
        }
    }

    @Override
    public void loadProducts(ArrayList<Product> products) {
        this.adapter.loadData(products);
    }

    @Override
    public void addProducts(ArrayList<Product> products) {
        this.adapter.addProducts(products);
    }

    @Override
    public void onClick(View view) {
        // tombol ubah ke tiles mode
        if (view == this.binding.ivList) {
            this.changePage(Page.PAGE_LIST_MODE);
        }
        // tombol show more
        else if (view == this.binding.showMore) {
            this.loadMoreProducts();
        }
    }

    //METHOD GANTI HALAMAN
    private void changePage(int page){
        Bundle result = new Bundle();
        result.putInt("page", page);
        this.fragmentManager.setFragmentResult("CHANGE_PAGE", result);
    }

}

File: app/src/main/java/com/example/proif_a01_java/View/ProductTilesFragments.kt
package com.example.proif_a01_java.View

import com.example.proif_a01_java.View.IProducts
import android.widget.AdapterView
import com.example.proif_a01_java.Model.Inventory
import com.example.proif_a01_java.Presenter.ProductPresenter
import com.example.proif_a01_java.View.ProductTilesAdapter
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.text.TextWatcher
import android.text.Editable
import android.view.View
import android.widget.Spinner
import android.widget.ArrayAdapter
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import com.example.proif_a01_java.Model.Page
import com.example.proif_a01_java.R
import com.example.proif_a01_java.Model.Product
import org.parceler.Parcels
import com.example.proif_a01_java.View.ProductTilesFragments
import com.example.proif_a01_java.databinding.FragmentProductTilesBinding
import java.util.ArrayList

class ProductTilesFragments  // constructor kosong
    : Fragment(), IProducts, View.OnClickListener, AdapterView.OnItemSelectedListener {
    lateinit var inv: Inventory
    lateinit var binding: FragmentProductTilesBinding
    lateinit var presenter: ProductPresenter
    lateinit var adapter: ProductTilesAdapter
    private var availLoads = 0
    private var loadNumber = 0
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentProductTilesBinding.inflate(inflater, container, false)

        // inisiasi
        presenter = ProductPresenter(this)
        adapter = ProductTilesAdapter(requireActivity(), presenter)

        // set adapter
        binding.lvProducts.adapter = adapter

        // filtering
        binding.etFilter.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                val search = binding.etFilter.text.toString()
                presenter.searchProducts(search)
            }
        })

        // spinner
        val ctSpinner = binding.categorySpinner
        val arr = arrayOf<String>("ALL", "MOBILE", "TABLET")
        val adapter: ArrayAdapter<*> =
            ArrayAdapter<Any>(requireContext(), R.layout.category_spinner_item, arr)
        adapter.setDropDownViewResource(R.layout.category_spinner_item)
        ctSpinner.adapter = adapter
        ctSpinner.onItemSelectedListener = this

        // set onclick listener
        binding.ivList.setOnClickListener { view: View -> onClick(view) }
        binding.showMore.setOnClickListener { view: View -> onClick(view) }

        // load new products
        this.loadProducts()

        // set adapter
        binding.lvProducts.adapter = this.adapter
        return binding.root
    }

    private fun loadProducts() {
        val products = ArrayList<Product>()
        for (i in 0..4) {
            products.add(inv.products[i])
        }
        loadNumber = 5
        availLoads = inv.products.size - 5
        presenter.loadProducts(products)
    }

    private fun loadMoreProducts() {
        if (availLoads > 0) {
            val products = ArrayList<Product>()
            val max: Int
            max = if (availLoads < 5) {
                loadNumber + availLoads
            } else {
                loadNumber + 5
            }
            for (i in loadNumber until max) {
                println(i)
                products.add(inv.products[i])
            }
            presenter.loadProducts(products)
            loadNumber = max
            availLoads -= 5
        }
    }

    override fun loadProducts(products: ArrayList<Product>) {
        adapter.loadData(products)
    }

    override fun moveToDetails(product: Product) {
        val result = Bundle()
        result.putParcelable("products", Parcels.wrap(product))
        parentFragmentManager.setFragmentResult("MOVE_DETAILS", result)
        changePage(Page.PAGE_DETAILS_MODE)
    }

    override fun onClick(view: View) {
        // tombol ubah ke tiles mode
        if (view === binding.ivList) {
            changePage(Page.PAGE_LIST_MODE)
        } else if (view === binding.showMore) {
            loadMoreProducts()
        }
    }

    //METHOD GANTI HALAMAN
    private fun changePage(page: Int) {
        val result = Bundle()
        result.putInt("page", page)
        parentFragmentManager.setFragmentResult("CHANGE_PAGE", result)
    }

    override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        presenter.changeCategory(i - 1)
    }

    override fun onNothingSelected(adapterView: AdapterView<*>) {}

    companion object {
        // singleton
        fun newInstance(inv: Inventory): ProductTilesFragments {
            val fragments = ProductTilesFragments()
            fragments.inv = inv
            return fragments
        }
    }
}